### PR TITLE
created coupons

### LINKED ISSUE #82: Feature: Seller Should Be Able to Create Discount Coupons
### Feature: Seller Should Be Able to Create Discount Coupons

**Description:**
We need to implement a feature that allows sellers to create and manage discount coupons. This involves backend support for creating, storing, and applying these coupons during the checkout process. The feature should include:

1. **Coupon Creation:**
   - Endpoint for sellers to create discount coupons.
   - Validation to ensure coupons have unique codes.
   - Fields to include: coupon code, discount percentage/amount, expiration date and applicable  product categories.

2. **Coupon Management:**
   - Endpoint for sellers to view, update, and delete coupons.
   - Proper validation and error handling.
   - Integrating coupons in checkout functions

3. **Database Integration:**
   - Schema design to store coupon information.
   - Migrations to create necessary database tables.

4. **Applying Coupons at Checkout:**
   - Logic to apply valid coupons to eligible products during the checkout process.
   - Ensure that discounts are calculated accurately.
   - Handle edge cases (e.g., expired coupons, invalid coupon codes).

**Tasks:**
- [x] Create endpoint for coupon creation.
- [x] Implement validation for unique coupon codes.
- [x] Design and implement database schema for storing coupon information.
- [x] Develop endpoint for managing (view, update, delete) coupons.
- [x] Implement logic to apply coupons at checkout.
- [x] Write unit and integration tests for all new functionality.
- [x] Update API documentation to include new endpoints and their usage.

**Acceptance Criteria:**
- Sellers can create discount coupons with the necessary fields.
- Coupons are stored in the database with unique codes.
- Sellers can manage their coupons (view, update, delete).
- Coupons can be applied during checkout, providing the correct discount to eligible products.
- All new functionality is tested and documented.

**Additional Notes:**
- Existing database will be used.
- Follow best practices for security and data validation.

### PR DESCRIPTION
**Pull request by niyobern**
#### What does this PR do?
This PR introduces new endpoints for managing coupons in our application. It provides the ability to create (POST), retrieve (GET), update (PUT), and delete (DELETE) coupons. Additionally, it includes a function that applies a coupon discount when called at checkout.

#### Description of Tasks completed
- Created a POST endpoint for adding new coupons by seller
- Created a GET all endpoint for retrieving existing coupons
- Created a GET one endpoint for retrieving a single coupon by id.
- Created a PUT endpoint for updating existing coupons
- Created a DELETE endpoint for removing coupons
- Implemented a function that applies a coupon discount when called in checkout functions.

#### How should this be manually tested?
To test the new endpoints:
1. Start the server
2. Open Swagger docs and navigate to tag "Coupon management" where you find endpoints for Coupon:
   - You need to authenticate with JWT
   - You get documentation of request body schema which you need to follow
   - Response schema is also documented. Check for discrepancies.
   - You need required permissions for each endpoint
   - You can't create coupons for products you don't own
   - You can't update/delete coupons you didn't create
   - Everything should behave as specified above.
3. To test the checkout discount function, you will need to import it and use it in your checkout function. Follow its doc strings to ensure everything works as expected.

### API Docs Screenshot
image

### Tests Coverage Screenshot
image

### CODE REVIEW

File: src/routes/couponRoute.ts
Comment (jkarenzi) [inline src/routes/couponRoute.ts:1]: Are these coupon routes to be accessed by both buyers and sellers? If not, i'd suggest checking the role before allowing access.

File: src/controller/couponController.ts [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+const createCouponRules = [
+    check('description').isLength({ min: 1 }).withMessage('coupon description is required'),
+    check('percentage').isInt({ min: 0, max: 100 }).withMessage('percentage must be between 0 and 100'),
+    check('expirationDate').isDate().withMessage('expiration date must be a valid date'),
+    check('applicableProducts').isArray().withMessage('applicable products must be an array of product ids'),
+]
+class CouponController {
+    // GET /coupons
+    public getAllCoupons = errorHandler(async (req: Request, res: Response) => {
```
Comment (jkarenzi) [inline src/controller/couponController.ts:28]: This function seems to get all coupons. Shouldn't you maybe have another function that retrieves the coupons that belong to a specific seller? and not all coupons?

File: src/controller/couponController.ts [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+                products.push(product);
+            }
+            coupon.applicableProducts = products;
+            const updatedCoupon = await couponRepository.save(coupon)
+            return res.status(200).json(updatedCoupon);
+        })
+    ]
+
+    // DELETE /coupons/:id
+    public deleteCoupon = errorHandler(async(req: Request, res: Response) => {
```
Comment (jkarenzi) [inline src/controller/couponController.ts:140]: Shouldn't you check to see if the coupon to be deleted actually belongs to the currently logged in user?

that way, a seller can only delete coupons that belong to them

**Review by jkarenzi — CHANGES_REQUESTED** [review]

File: src/routes/couponRoute.ts
Comment (niyobern) [reply src/routes/couponRoute.ts:1]: coupon post route check that you are creating coupons for your own products (buyers don't own products); put route also check that you are modifying your own coupon. for delete, I should also be so but I'm sorry. It is going to be fixed by the next 44 minutes.

File: src/controller/couponController.ts (hunk 1 above)
Comment (niyobern) [reply src/controller/couponController.ts:28]: I'm currently solving that and in 26 minutes, I will be done

Comment (niyobern) [reply src/controller/couponController.ts:28]: That has been done

File: src/controller/couponController.ts (hunk 2 above)
Comment (niyobern) [reply src/controller/couponController.ts:140]: It's done

File: src/routes/couponRoute.ts
Comment (niyobern) [reply src/routes/couponRoute.ts:1]: Buyers have read-only permission on coupons. It's done.

**Review by ambroisegithub — APPROVED** [review]
It Looks Good for merge!!!

**Review by EddyShimwa — APPROVED** [review]
👍

**Review by jkarenzi — APPROVED** [review]

**Review by dushimeemma — APPROVED** [review]
Good job 🥇 🥇

**Pull request merged**